src/deluge/dsp/oscillators/sine_osc.h

```diff
@@ -25,45 +25,38 @@ class SineOsc {
 		return value + diff * strength2;
 	}
 
-	static inline int32x4_t getSineVector(uint32_t* thisPhase, uint32_t phaseIncrement) {
+	// Test and verified to be identical to original function 11/22/24
+	static inline Argon<int32_t> getSineVector(uint32_t* thisPhase, uint32_t phaseIncrement) {
+		// create a vector containing a ramp of incrementing phases:
+		// {phase + (phaseIncrement * 1), phase + (phaseIncrement * 2), phase + (phaseIncrement * 3), phase +
+		// (phaseIncrement * 4)}
+		auto phaseVector = Argon<uint32_t>::LoadCopy(thisPhase).MultiplyAdd(uint32x4_t{1, 2, 3, 4}, phaseIncrement);
+		*thisPhase = *thisPhase + phaseIncrement * 4;
 
-		int16x4_t strength2;
-		uint32x4_t readValue;
-
-		for (int32_t i = 0; i < 4; i++) {
-			*thisPhase += phaseIncrement;
-			int32_t whichValue = *thisPhase >> (32 - kSineTableSizeMagnitude);
-			strength2[i] = (*thisPhase >> (32 - 16 - kSineTableSizeMagnitude + 1)) & 32767;
-
-			uint32_t readOffset = whichValue << 1;
-
-			readValue[i] = *(uint32_t*)&sineWaveDiff[readOffset];
-		}
-
-		int32x4_t enlargedValue1 = vreinterpretq_s32_u32(vshlq_n_u32(readValue, 16));
-		int16x4_t diffValue = vshrn_n_s32(vreinterpretq_s32_u32(readValue), 16);
-
-		return vqdmlal_s16(enlargedValue1, strength2, diffValue);
+		return render(phaseVector);
 	}
 
-	static inline int32x4_t doFMVector(uint32x4_t phaseVector, uint32x4_t phaseShift) {
-
-		uint32x4_t finalPhase = vaddq_u32(phaseVector, vshlq_n_u32(phaseShift, 8));
+	// Test and verified to be identical to original function 11/22/24
+	static inline Argon<int32_t> doFMVector(Argon<uint32_t> phaseVector, Argon<uint32_t> phaseShift) {
+		return render(phaseVector + (phaseShift << 8));
+	}
 
-		uint32x4_t readValue{0};
-		util::constexpr_for<0, 4, 1>([&]<int i>() {
-			uint32_t readOffsetNow = (vgetq_lane_u32(finalPhase, i) >> (32 - kSineTableSizeMagnitude)) << 1;
-			uint32_t* thisReadAddress = (uint32_t*)&sineWaveDiff[readOffsetNow];
-			readValue = vld1q_lane_u32(thisReadAddress, readValue, i);
-		});
+private:
+	static inline Argon<int32_t> render(Argon<uint32_t> phase) {
+		// interpolation fractional component
+		ArgonHalf<int16_t> strength2 = phase.ShiftRightNarrow<32 - 16 - kSineTableSizeMagnitude + 1>()
+		                                   .BitwiseAnd(std::numeric_limits<int16_t>::max())
+		                                   .As<int16_t>();
 
-		int16x4_t strength2 =
-		    vreinterpret_s16_u16(vshr_n_u16(vshrn_n_u32(finalPhase, (32 - 16 - kSineTableSizeMagnitude)), 1));
+		// multiply by 2 due to interleave. i.e. sin(x) is actually table[x * 2]
+		Argon<uint32_t> indices = (phase >> (32 - kSineTableSizeMagnitude)) << 1;
 
-		int32x4_t enlargedValue1 = vreinterpretq_s32_u32(vshlq_n_u32(readValue, 16));
-		int16x4_t diffValue = vshrn_n_s32(vreinterpretq_s32_u32(readValue), 16);
+		//  load our two relevent table components
+		auto [sine, diff] = ArgonHalf<int16_t>::LoadGatherInterleaved<2>(sineWaveDiff.data(), indices);
```

Comment:
This isn't a 1:1 refactor, but maybe the use of load2s makes up for the extra neon->CPU transfer for the phase vector? Otherwise LGTM

Reply:
Under the hood it uses the same `vld1q_lane_s32` intrinsic (tho in this case due to deinterleaving it's actually `vld2q_lane_s16`)